src/storage/mvcc/reader.rs

```diff
@@ -417,4 +418,236 @@ impl<'a> MvccReader<'a> {
             keys.push(key);
         }
     }
+
+    // Returns true if it needs gc.
+    // This is for optimization purpose, does not mean to be accurate.
+    pub fn need_gc(&self, safe_point: u64) -> bool {
+        let mut opts = GetPropertiesOptions::default();
+        opts.max_ts = Some(safe_point);
+        let props = match self.snapshot.get_properties_cf(CF_WRITE, &opts) {
+            Ok(v) => v,
+            Err(_) => return true,
+        };
+        // We don't need gc if:
+        // 1. min_ts > safe_point:
+        //    No data older than safe_point to gc.
+        if props.min_ts > safe_point {
+            return false;
+        }
+        // 2. num_rows == num_versions && num_puts == num_versions:
+        //    No rows have more than one versions, and the version is effective.
+        //    Notice: Since the properties are file-based, it can be false positive.
+        //    For example, if multiple files have a different version of the same row.
+        !(props.num_rows == props.num_versions && props.num_puts == props.num_versions)
```

Comment:
2) and 3) are covered by 1) in a way:
2) `num_mvcc_del` is included in `num_versions`
3) If a single row has too many versions, the total `num_versions` may or may not > `num_rows * threshold`. Do you want to set a fixed number for this case, like 10 or 100 versions of a single row?